demos/python_demos/formula_recognition_demo/utils.py

```diff
@@ -1,11 +1,339 @@
+            return None
+
+        if self.model_status == Model.Status.encoder_infer:
+            infer_status_encoder = self._infer_request_handle_encoder.wait(timeout=0)
+            if infer_status_encoder == 0:
+                self._run_decoder()
+            return None
+
+        return self._process_decoding_results()
+
+    def infer_sync(self, model_input):
+        assert not self.is_async
+        model_input = change_layout(model_input)
+        self._run_encoder(model_input)
+        self._run_decoder()
+        res = None
+        while res is None:
+            res = self._process_decoding_results()
+        return res
+
+    def _process_decoding_results(self):
+        timeout = 0 if self.is_async else -1
+        infer_status_decoder = self._infer_request_handle_decoder.wait(timeout)
+        if infer_status_decoder != 0 and self.is_async:
+            return None
+        dec_res = self._infer_request_handle_decoder.output_blobs
+        self._unpack_dec_results(dec_res)
+
+        if self.tgt[0][0][0] == END_TOKEN or self.num_infers_decoder >= self.args.max_formula_len:
+            self.num_infers_decoder = 0
+            self.logits = np.array(self.logits)
+            logits = self.logits.squeeze(axis=1)
+            targets = np.argmax(logits, axis=1)
+            self.model_status = Model.Status.ready
+            return logits, targets
+        self._infer_request_handle_decoder = self._async_infer_decoder(self.row_enc_out,
+                                                                       self.dec_states_c,
+                                                                       self.dec_states_h,
+                                                                       self.output,
+                                                                       self.tgt,
+                                                                       req_id=0
+                                                                       )
+
+        return None
+
+    def _run_encoder(self, model_input):
+        timeout = 0 if self.is_async else -1
+        self._infer_request_handle_encoder = self._async_infer_encoder(model_input, req_id=0)
+        self.model_status = Model.Status.encoder_infer
+
+    def _run_decoder(self):
+        enc_res = self._infer_request_handle_encoder.output_blobs
+        self._unpack_enc_results(enc_res)
+        self._infer_request_handle_decoder = self._async_infer_decoder(
+            self.row_enc_out, self.dec_states_c, self.dec_states_h, self.output, self.tgt, req_id=0)
+        self.model_status = Model.Status.decoder_infer
+
+    def _unpack_dec_results(self, dec_res):
+        self.dec_states_h = dec_res[self.args.dec_st_h_t_layer].buffer
+        self.dec_states_c = dec_res[self.args.dec_st_c_t_layer].buffer
+        self.output = dec_res[self.args.output_layer].buffer
+        logit = dec_res[self.args.logit_layer].buffer
+        self.logits.append(logit)
```

Comment:
This is the only field which stores a list of outputs. Just store the output like others. This will remove the need in
```py
            self.logits = np.array(self.logits)
            logits = self.logits.squeeze(axis=1)
```
in `_process_decoding_results()`.